File: src/jitter_analysis/src/mock_robot_driver.cpp
#include <chrono>
#include <memory>
#include <random>
#include <thread>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/joint_state.hpp"

using namespace std::chrono_literals;

class MockRobotDriver : public rclcpp::Node
{
public:
  MockRobotDriver()
  : Node("mock_robot_driver")
  {
    publisher_ = this->create_publisher<sensor_msgs::msg::JointState>("/joint_states", 10);
    timer_ = this->create_wall_timer(8ms, std::bind(&MockRobotDriver::timer_callback, this));
    RCLCPP_INFO(this->get_logger(), "模拟机器人驱动已启动，正在以 125Hz 发布数据...");
  }

private:
  void timer_callback()
  {
    auto msg = sensor_msgs::msg::JointState();
    msg.header.stamp = this->now();
    msg.name = {"joint_1", "joint_2", "joint_3", "joint_4", "joint_5", "joint_6"};
    msg.position = {0.0, 0.0, 0.0, 0.0, 0.0, 0.0};
    
    // --- 模拟系统抖动 (可选) ---
    // 如果你想模拟真实系统的抖动，可以取消下面这一行的注释：
    // std::this_thread::sleep_for(std::chrono::microseconds(std::rand() % 2000)); 
    
    publisher_->publish(msg);
  }
  
  rclcpp::Publisher<sensor_msgs::msg::JointState>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<MockRobotDriver>());
  rclcpp::shutdown();
  return 0;
}

File: run/test_python/mock_robot_driver.py
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import JointState
import time
import random

class MockRobotDriver(Node):
    def __init__(self):
        super().__init__('mock_robot_driver')
        self.publisher_ = self.create_publisher(JointState, '/joint_states', 10)
        
        # 目标频率 125Hz (8ms)
        self.timer_period = 0.008  
        self.timer = self.create_timer(self.timer_period, self.timer_callback)
        
        self.get_logger().info('模拟机器人驱动已启动，正在以 125Hz 发布数据...')

    def timer_callback(self):
        msg = JointState()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.name = ['joint_1', 'joint_2', 'joint_3', 'joint_4', 'joint_5', 'joint_6']
        msg.position = [0.0] * 6
        
        # --- 模拟系统抖动 (可选) ---
        # 如果你想模拟真实系统的抖动，可以取消下面这一行的注释：
        # time.sleep(random.uniform(0, 0.002)) 
        
        self.publisher_.publish(msg)

def main():
    rclpy.init()
    node = MockRobotDriver()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == '__main__':
    main()